File: lintcode/rotate-list.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/rotate-list/

from linked_list import *

class Solution:
    # @param head: the list
    # @param k: rotate to the right k places
    # @return: the list after rotation
    def rotateRight(self, head, k):
        # write your code here
        count = 0
        p = head
        tail = None
        while p:
            count += 1
            tail = p
            p = p.next
        if count <= 1:
            return head

        k = k % count
        if k == 0:
            return head

        p = None
        i = 0
        while i < count - k:
            if not p:
                p = head
            else:
                p = p.next
            i += 1

        tail.next = head
        head = p.next
        p.next = None
        return head

File: leetcode-2/rotate-list.cc
// Question: https://leetcode.com/problems/rotate-list/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || !head->next || k == 0) {
            return head;
        }

        auto node = head;
        int count = 0;
        ListNode* tail = NULL;
        while (node) {
            ++count;
            tail = node;
            node = node->next;
        }

        if (k % count == 0) {
            return head;
        }

        int n = count - k % count;
        node = head;
        while (--n) {
            node = node->next;
        }

        auto ans = node->next;
        node->next = NULL;
        tail->next = head;
        return ans;
    }
};
